privacyidea/lib/eventhandler/customuserattributeshandler.py

```diff
@@ -0,0 +1,87 @@
+__doc__ = """This is the event handler module for custom user attributes.
+You can set or delete custom your attributes
+
+"""
+
+from privacyidea.lib.eventhandler.base import BaseEventHandler
+from privacyidea.lib.user import User
+from privacyidea.lib import _
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class VALIDITY(object):
+    """
+    Allowed validity options
+    """
+    START = "valid from"
+    END = "valid till"
+
+
+class CustomUserAttributesHandler(BaseEventHandler):
+    """
+    An Eventhandler needs to return a list of actions, which it can handle.
+
+    It also returns a list of allowed action and conditions
+
+    It returns an identifier, which can be used in the eventhandlig definitions
+    """
+
+    identifier = "CustomUserAttributes"
+    description = "This event handler can set and delete custom_user_attributes"
+
+    @property
+    def allowed_positions(cls):
+        """
+        This returns the allowed positions of the event handler definition.
+        :return: list of allowed positions
+        """
+        return ["post", "pre"]
+
+    @property
+    def actions(cls):
+        """
+        This method returns a dictionary of allowed actions and possible
+        options in this handler module.
+
+        :return: dict with actions
+        """
+        actions = {
+            "set_custom_user_attributes":{
+                "attrkey": {
+                    'type': 'str',
+                    'discription': _('The key of the attribute')},
+                "attrvalue": {
+                    'type': 'str',
+                    'discription': _('The value of the attribute')}
+            },
+            "delete_custom_user_attributes":{
+                "user": {
+                    'type': 'user',
+                    'discription': _('An object from the class user')}
+        }}
+        return actions
+
+    def do(self, action, options=None):
+        """
+        This method executes the defined action in the given event.
+
+        :param action:
+        :param options: Contains the flask parameters g,  attrkey and attrvalue
+        :type options: dict
+        :return:
+        """
+        ret = True
+        g = options.get("g")
+        attrkey = options.get("attrkey")
+        attrvalue = options.get("attrvalue")
+        logged_in_user = g.logged_in_user
```

Comment:
I see we have been a bit unspecific here in the definition of the issue.
I think setting the attribute of the logged in user does very seldomly make sense.

I guess often we might want to set a custom attribute of a user, who authenticates at /validate/check or if an administrator acts on a user or on the token of a user.

So determining "logged_in_user" is not enough.

You should use the user from the request.
OR: we might even define in the event handler on which user we want to act: either on the "logged_in_user" or on the "tokenowner"; similar to the notification handler:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/eventhandler/usernotification.py#L177

So the `actions` would be sth like:

~~~~python
"set_custom_user_attributes":{
                "user" : {
                    'type': 'str',
                    'required': True,
                    'value': ["logged in user", "tokenowner"] },
                "attrkey": {
                    'type': 'str',
                    'required': True,   
                    'description': _('The key of the attribute')},
                "attrvalue": {
                    'type': 'str',
                    'required': True,
                    'description': _('The value of the attribute')}
            },
  "delete_custom_user_attributes":{
      "user": {
                    'type': 'str',
                    'required': True,
                    'value': ["logged in user", "tokenowner"] },
        "attrkey": {
            'type': 'str',
            'required': True,   
            'description': _('The custom attribute of the user to delete.')},
~~~~

Please check the user notification how to determine the user object either from the logged in user or from the tokenowner.

Reply:
https://github.com/privacyidea/privacyidea/compare/%232852/Add_event_handler_for_custom_user_attributes